live_data_subscriber/mqtt/__init__.py

```diff
@@ -22,14 +22,15 @@
 }
 
 
-def generate_topic_api_client_args_mapping():
+def generate_api_client_args_mapping():
     sim_api_domain_name = environ["API_CLIENT_SIM_API_DOMAIN_NAME"]
     external_api_domain_name = environ["API_CLIENT_DOMAIN_NAME"]
     websocket_api_domain_name = environ["API_CLIENT_WEBSOCKET_DOMAIN_NAME"]
 
     cn_mapping = list_running_canary_networks_and_devices_with_live_data(sim_api_domain_name)
```

Comment:
If the token authentication fails or there are no Canary Networks available, the script will crash at this point because of this line: https://github.com/gridsingularity/d3a-api-client/blob/7116667243cb34097d3a63855219f72b72c5f0cc/d3a_api_client/utils.py#L214

I think it would be good to add a check in `list_running_canary_networks_and_devices_with_live_data` to make sure that:
- `data` is returned (and it's not None)
- there have been no errors during the request

Reply:
On this line: https://github.com/gridsingularity/d3a-api-client/blob/7116667243cb34097d3a63855219f72b72c5f0cc/live_data_subscriber/mqtt/__init__.py#L37

if the `cn_mapping` is empty (when there are no CNs) the loop won't do anything, but the method will still return `topic_api_client_mapping` and `device_api_client_mapping`. So the program will keep running but it will fail later on. Do you think this should be handled somehow, or is this expected?

Reply:
> If the token authentication fails , the script will crash at this point because of this line
To me it sounds like a good thing, as authentication must correctly done to continue reading data from energy meters and to be forwarded to respective CNs.

> If there are no Canary Networks available, the script will crash at this point because of this line
It didn't crashed for me, only that the received data from respective energy meter simply isn't forwarded as the mapping is empty due to no CN

Reply:
> the program will keep running but it will fail later on. Do you think this should be handled somehow, or is this expected?

Thanks for prolonged testing, the culprit was at this [line](https://github.com/gridsingularity/d3a-api-client/pull/113/files/6cfe04baa216f6f066e4f4d2ace87bb3dd3f4b04?file-filters%5B%5D=.py&hide-deleted-files=true#diff-2ac8c9997c6823e9284fe32ec6bfe4666ea735e13ce528bd7be42b6f7aa094d0R43), when after some some interval CN's device mapping is refreshed in order to take into account the newly added CN, earlier it was only updating the mapping for `mqtt` but after incorporating for `ws` the returned format was changed, that failed the script later, applied a fix via this [commit](https://github.com/gridsingularity/d3a-api-client/pull/113/commits/6c5cdefbd36b131b72409de614f84d7573f19987). Feel free to give another try.

Reply:
> To me it sounds like a good thing, as authentication must be correctly done to continue reading data from energy meters and to be forwarded to respective CNs.

I agree with you that the program should stop. I think it would be better if it would exit gracefully rather than crashing, though. Do you think we could show the user a message saying that something went wrong with the authentication? This would probably involve changing this part: https://github.com/gridsingularity/d3a-api-client/blob/6c5cdefbd36b131b72409de614f84d7573f19987/d3a_api_client/utils.py#L60-L64

Rather than returning `None`, we could raise an exception and then handle it where the method is called:
```python
try:
    data = execute_graphql_request(domain_name=domain_name, query=query)
except SomeCustomAuthenticationException as ex:
    # Exit gracefully with a message
    sys.exit(ex)  # The message might need to be written differently
```

What do you think?

Reply:
Im not sure if gracefully shutting would help k8s to restart the service on its own. @spyrostz @hannesdiedrich ?

Reply:
@faizan2590 no matter how the service is terminated the service will be restarted by K8s, but I see little help in that regard, since IIRC sys.exit also raises an exception `SystemExit` , therefore the application will still terminate with an unhandled exception, and I suspect that the annoying traceback will still be present. IMO the only advantage in terminating gracefully in this case is to mute the traceback from an exception that you do not want to see. 

Give it a try, but if the traceback is still reported in this case I would personally prefer to not use sys.exit at all, because there is no gain over what we have now.

Reply:
I think I might be missing some part of the broader view for this PR: since I am running the script locally as described in the description of the PR, I'm not aware of the role of k8s. Exiting gracefully would only make sense if this script was used by a user that might need some kind of and readable output. If it's part of an automated pipeline, this is not so important I guess :)

Reply:
This is essentially the subscriber for the data reported for Simris. Running the script locally is mainly for testing purposes, and not by users other than us. The main purpose of this script is to live in our cluster, connect to the Simris server via websocket to read the measurements for their devices, and propagate these measurements to d3a.

Reply:
`retrieve_jwt_key_from_server ` is being used in many other places, if I change it now from returning `None` to raising an exception other parts of codes has to be adapted too. I would propose to create an small ticket for that as it seems to deviates from the main purpose of this ticket.